Automattic/wp-calypso
# Add A/B test to move user step to last position of `onboarding` flow

This is a 50/50 A/B test to test out moving the user / account step to the end of the onboarding flow. Currently users have to enter their account details before progressing through the flow. This change would move it to the end, similar to how the signup flow used to be.

***Some caveats:***

A Happychat error is thrown in the console on the `plans` step because `getHappychatAuth` fails since there is no user currently logged in. The impact to the user is that the inline help button (question mark in the bottom right hand corner) isn't there throughout the signup flow. Again, this might be okay for the purposes of the test, but if we permanently move the `user` step, we might want to look into a way to keep it for logged out users.

#### Changes proposed in this Pull Request

* Create A/B test `moveUserStepPosition` and conditionally swap the defaultFlowName from `onboarding` to `onboarding-user-last`.
* When the user is in the `last` group, also add an additional flow that is a duplicate of the `ecommerce-onboarding` flow, but with the `user` step in the last position. This allows switching between the onboarding and ecommerce flows while keeping the user step last, and avoiding the known issue with broken transitions between flows: https://github.com/Automattic/wp-calypso/issues/32506
* Update the Signup component to _not_ render the step if in the process of performing a redirect. This avoids a BSOD issue when the `user` step renders and calls `saveSignupStep`, potentially putting the `user` step in the wrong position in the progress state.
* Update the login and initial-state behaviour to _not_ remove persisted logged out state, and to copy it to the current state if there is no existing persisted state for a logged in user, but there _is_ persisted state for a logged out user. This will allow a user to complete the signup process, get to the last step, sign in, and resume their progress from the last step instead of having to enter everything in again. The intention is to effectively reinstate this bug fix from last year, but now in Redux: https://github.com/Automattic/wp-calypso/pull/25690
* Don't hide the login link in the footer of the `user` step when the step is in a position other than 0

#### Demo of logging in at the final step and being redirected back to the plans step

![move-user-step-last-ab-test-small](https://user-images.githubusercontent.com/14988353/61854181-e3e7f080-af00-11e9-8208-748bb9efe4c5.gif)

#### Testing instructions



* Add the `no-force-sympathy` config flag and set it to `true`, either in your `development.json` file, or as an environment variable before you run `npm run start`. This will switch off the "Sympathy" feature in your local development environment, which randomly clears the state on page load 25% of the time. Switching this off will ensure that the login redirect will work correctly.

##### Test variant

* In the Calypso dev menu at the bottom right of the screen, set the `moveUserStepPosition` value to `last`
* From logged out, go to `/start/` and it should take you immediately to the `site-type` step. Complete the flow with blog / business / or professional site types. The final step should be the `user` step. After creating an account, it should immediately take you to the loading screen for creating a new site, and then on to the checklist.
* Repeat the steps above but select the `online store` site type, and this should take you to `/start/online-store-onboarding/domains`

##### Test variant with logging in

Note: the post login redirect will not resume progress to the previous step on a Private browser in Firefox due to it not persisting IndexedDB beyond a page reload / redirect. However, Chrome Incognito browser will work (and normal Firefox should work fine, too).

* In the Calypso dev menu at the bottom right of the screen, set the `moveUserStepPosition` value to `last`
* From logged out, go to `/start/` and it should take you immediately to the `site-type` step. Complete the flow with blog / business / or professional site types. The final step should be the `user` step.
* Log in to an existing account, and it should redirect you to the final step (plans).
* Select a plan (e.g. Start with a free site) and you should be redirected to the checklist page of the new site.
* Repeat the steps above but select the `online store` site type, and this should take you to `/start/online-store-onboarding/domains`.

##### Test control

* In the Calypso dev menu at the bottom right of the screen, set the `moveUserStepPosition` value to `first`
* From logged out, go to `/start/` and it should take you to the user step
* Log in, and from the `site-type` step, select `online-store`. This should take you to `/start/ecommerce-onboarding/domains`
* Completing signup should be as it was on master before applying this AB test
